### ISSUE TITLE
2 parallel GPUs and PyTorch from source

### ISSUE THREAD

**Issue by jmariscal-harana** [MIC-DKFZ/nnUNet#367]
Hi Fabian,

First of all, thanks so much for the amazing framework and for the commented code!
I am new to cardiac segmentation and I am struggling with 3 things which I would really appreciate your help with. Just for context, these are some of my system specs and how I installed nnUNet and PyTorch:

- Ubuntu 18.04, threadripper 3960x, 2x TITAN RTX, 1TB SSD, 2x64GB DDR4
- CUDA 11.1, cuDNN 8.0.4 (torch.backends.cudnn.version()), PyTorch compiled from source, nnUNet installed as an integrative framework within virtualenv (python 3.6.9)

I first tried both examples (train and test) and they worked nicely for me. I then tried to train the 2D trainer using my 2 GPUs and my own data:
- nnUNet_plan_and_preprocess -t XXX --verify_dataset_integrity (worked well)
- CUDA_VISIBLE_DEVICES=0,1 python -m torch.distributed.launch --master_port=4321 --nproc_per_node=2 run/run_training_DDP.py 2d nnUNetTrainerV2_DDP XXX all --dbs --fp32

With this config, each epoch took around 50s (I added --fp32 after my first run crashed at epoch 198 with OOM). The dice validation results were nice (0.98/0.99). I trained for 1000 epochs but didn't manage to use nnUNet_change_trainer_class properly to allow me to run inferences and evaluate model performance on my test set. So my 3 questions are:

1. What arguments should I be adding to nnUNet_change_trainer_class so that I can run inferences on my test set and evaluate?
2. I am now training with a single GPU (nnUNet_train 2d nnUNetTrainerV2 XXX all --fp32), but each epoch takes 147s (nvidia-smi shows that one GPU is using 16GB and the other one is idle). Does this make sense? 
3. We have quite a large dataset and we have a train/val/test split which we would like to use. So far I am ignoring the val folder and training with the train folder only (hence fold=all), but ideally I would like to use train/val properly during training. Is there a way to do this?

Thanks a lot for all the work and for sharing it, Fabian!

Kind regards,
Jorge
PS: apologies for the long post and if some questions are basic; I tried reading all the docs and searching previous issues before posting!

**Comment by FabianIsensee** [comment]
Hi,

that's a nice system. Finally someone who does not underspec their CPU ;-)

Regarding your questions:
1) If you run `nnUNet_change_trainer_class -h` it will tell you what you need to know.
2) I personally always use single GPU training. Scaling is better to run multiple trainings in parallel (if I run 5 folds on 5 GPUs it is faster to run 1 fold per GPU rather than sequentially run the folds with a 5 GPU DDP training). 
3) You need to merge your train and val data. Put them all into the imagesTr and labelsTr folders. Then follow the instructions here: https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/common_questions.md#creating-and-managing-data-splits to generate your custom train:val split

I am confused that you are still running into OOM issues. I thought this was fixed. Building pytorch was done in an anaconda environment, right? Can you please run the following script and observe the RAM usage while this runs? Does the RAM increase with every epoch?

```python
import gc

from nnunet.network_architecture.generic_UNet import Generic_UNet
from nnunet.network_architecture.initialization import InitWeights_He
from nnunet.training.loss_functions.dice_loss import DC_and_CE_loss
from torch import nn
from torch.cuda.amp import autocast, GradScaler
from torch.optim import SGD
import numpy as np
import torch


def dummy_dl(batch_size, patch_size, data_channels, num_classes):
    data = np.random.random((batch_size, data_channels, *patch_size))
    seg = np.random.uniform(0, num_classes-1, (batch_size, 1, *patch_size)).round()
    data_torch = torch.from_numpy(data).float()
    seg_torch = torch.from_numpy(seg).float()
    while True:
        yield data_torch, seg_torch


def main():
    patch_size = [320, 256]
    batch_size = 40
    data_channels = 1
    num_classes = 2

    model = Generic_UNet(data_channels, 16, num_classes, 6, 2, 2, nn.Conv2d, nn.InstanceNorm2d,
                         {'eps': 1e-05, 'affine': True}, nn.Dropout2d,
                         {'p': 0, 'inplace': True}, nn.LeakyReLU, {'negative_slope': 0.01, 'inplace': True}, False,
                         False, lambda x: x, InitWeights_He(1e-2), [[2, 2], [2, 2], [2, 2], [2, 2], [2, 2], [2, 2]],
                         [[3, 3], [3, 3], [3, 3], [3, 3], [3, 3], [3, 3], [3, 3]], False, True, True, 512)
    model = model.cuda()
    optimizer = SGD(model.parameters(), 0.01, 0.99, nesterov=True, weight_decay=3e-5)

    dl = dummy_dl(batch_size, patch_size, data_channels, num_classes)
    dl_val = dummy_dl(batch_size, patch_size, data_channels, num_classes)

    loss_fn = DC_and_CE_loss({'batch_dice': True, 'smooth': 1e-5, 'do_bg': False}, {})

    scaler = GradScaler()

    """import hiddenlayer as hl
    g = hl.build_graph(model, torch.rand((1, 1, *patch_size)).cuda(), transforms=None)
    g.save("/home/fabian/network_architecture.pdf")"""

    for ep in range(1000):
        print(ep)
        losses_train = []
        for batch in range(30):
            optimizer.zero_grad()

            data, seg = next(dl)
            data = data.cuda()
            seg = seg.cuda()

            with autocast():
                output = model(data)
                loss = loss_fn(output, seg)

            scaler.scale(loss).backward()
            #loss.backward()

            scaler.step(optimizer)
            #optimizer.step()

            scaler.update()

            losses_train.append(loss.item())
        print(np.mean(losses_train))

        losses_val = []
        with torch.no_grad():
            for batch in range(30):
                data, seg = next(dl_val)
                data = data.cuda()
                seg = seg.cuda()

                with autocast(True):
                    output = model(data)
                    loss = loss_fn(output, seg)
                losses_val.append(loss.item())
            #gc.collect()
        print(np.mean(losses_val))


if __name__ == "__main__":
    main()
```

If yes, please try an anaconda environment with python 3.8.5. I use Ubuntu 18.04, 2080ti, python 3.8.5 and pytorch 1.6.0 with cuDNN 8.0.2 and do not have these issues (I did have them in the past though).

Best,
Fabian

**Comment by jmariscal-harana** [comment]
Hi,

Thanks for the quick and thorough reply. I was lucky enough to have the help of someone who understands deep learning bottlenecks better than I do for my specs :)

1./2. Given those facts, I am now training on a single GPU.
3. Fantastic, thanks!

My apologies, I just realised that the OOM issue happened before compiling PyTorch from source. I am using virtualenv (Ubuntu 18.04, Titan RTX, python 3.6.9, CUDA 11.1, cuDNN 8.0.4, pytorch 1.8.0), not conda :) After compiling, I don't get any OOM issues. I am now running one 2D model on each GPU with mixed precision, and RAM consumption is stable at 7GB, with 95% GPU-util and max power. Overall RAM usage is stable, too. Each epoch takes around 85 s, so by looking at the benchmarks I guess that all is good now.

Thanks a lot!

Kind regards,
Jorge

**Comment by FabianIsensee** [comment]
Dear Jorge,
great to hear everything has been resolved. And thank you very much for the feedback about the OOM issue.
Good luck with your trainings ;-)
Best,
Fabian
